src/loaders/FileLoader.js

```diff
@@ -239,7 +226,155 @@ class FileLoader extends Loader {
 
 		scope.manager.itemStart( url );
 
-		return;
+		return res;
+
+	}
+
+	async loadAsync( url ) {
```

Comment:
There's no need to implement `loadAsync` for FileLoader because it's implemented in the super class Loader here:

https://github.com/mrdoob/three.js/blob/8cb903d61618cb58e3134431a389212d9c98dff0/src/loaders/Loader.js#L19-L29

`FileLoader.loadAsync` will automatically return a promise that resolves with the value passed to `onLoad` in the load function.

Reply:
Yes, the resolve of the promise has the value. Agreed. But the `load` function is not `.then`able without destroying the promise because the `load` function doesn't return the value. I'll demonstrate:
The setup is two files. `main.js` and `data.txt`. `data.txt` has any placeholder values. `main.js` starts as follows:
```javascript
import { FileLoader } from 'https://cdn.skypack.dev/three';

const loader = new FileLoader()
const asset = loader.loadAsync("./data.txt")
console.log(asset);

setTimeout(()=> {
  console.log(asset);
}, 10)
```
The output of running this in chrome:
```
Promise {<pending>}
main.js:8 Promise {<fulfilled>: 'asdljfhbaboijshdjhbasdghoiasdf'}
```
Now, change `main.js` like so:
```javascript
import { FileLoader } from 'https://cdn.skypack.dev/three';

const loader = new FileLoader()
const asset = loader.loadAsync("./data.txt").then(val => {
  console.log(val);
})
console.log(asset);

setTimeout(()=> {
  console.log(asset);
}, 10)
```
The output is now:
```
Promise {<pending>}
main.js:5 asdljfhbaboijshdjhbasdghoiasdf
main.js:10 Promise {<fulfilled>: undefined}
```

Notice the `undefined`. This means anyone wanting to do the following pattern:
```javascript
const promise = loader.loadAsync().then(val => {
  // extra stuff to val
})

// do other stuff in the mean time

await promise
```
is completely unable to do so because the promise resolves early with `undefined`.

Reply:
This is just how promise chains work, though -- it doesn't have anything to do with the Loader or what the load function returns. The promise resolves to `undefined` in your second case because you've chained a "then" to the original one without returning anything from it:

```js
const promise = loader.loadAsync("./data.txt").then(val => {
  console.log(val); // text data

  // returns nothing so next "then" will get undefined
});

promise.then(val => {
  console.log(val); // undefined
});
```

If you want to chain your promises immediately and still have the loaded data in subsequent "then" callbacks you need to return it:

```js
const promise = loader.loadAsync("./data.txt").then(val => {
  console.log(val); // text data

  return val;
});
```

Or just keep a copy of the original promise so you can use that:

```js
const promise = loader.loadAsync("./data.txt");
promise.then(val => {
  console.log(val);
});

setTimeout(() => {
  promise.then(val =>{
    console.log(val); // still valid data and promise will say "fulfilled" with the text data
  });
}, 10);
```

Reply:
... *realizes the error*... 🤦‍♂️ thank you again @gkjohnson. Here's to learning.

Reply:
Yeah promises are not the most intuitive -- took me a bit to understand the details of them